### DISCUSSION TITLE
Can `required` lexicon fields have null or false-y values?

### DISCUSSION THREAD

**Discussion by snarfed** [bluesky-social/atproto#2684] [Protocol (atproto)]
Minor nitpicky question: https://atproto.com/specs/lexicon#object says:

> `required` (array of strings, optional): indicates which properties are required

So the field must be present. Can its value be `null`? Or [false-y](https://atproto.com/specs/data-model#nullable-and-false-y), eg the empty string?

(This came up recently in the API Discord 😁)

**Comment by snarfed** [comment]
@DavidBuchanan314 says this probably falls through to `nullable`. Makes sense if so!

**Reply by bnewbold** [reply]
If a field is `required`, the key must be present in the JSON/CBOR object.

If a field is `nullable`, it may have have `null` as the value.

We don't treat "false-y" as a thing anywhere, I think: `null` is distinct from field not existing, which is distinct from things like empty-string, 0, or empty array.

Note that there is also a `null` field type.

Per the specs today, it is allowed to be both not-`required` and `nullable`, but we really strongly recommend against this. There are one or two instances of this in the current Lexicons, and we would like to remove them and update the spec to say that a field can either be optional or not-`required` but not both. The main motivation for this is the Go JSON marshalling library in the Go standard library, which does not have a way of representing "double-optional": if the Go value is `null` it either needs to serialize as "field doesn't exist at all", or `null`, there isn't a way to select the behavior at runtime. This is mentioned in https://github.com/bluesky-social/atproto/discussions/2128 as "Nullable and Optional".
